### PR TITLE
HPCC-18900 Apply keyed filters on translated disk reads correctly

### PR DESCRIPTION
**Pull request by richardkchapman**
Completing the work in the prior commit in the area that was not 100%
finished.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Updated the test, and manually stepped through a few variants to ensure coverage was as intended. Should consider adding some miskeyed variants as well.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

File: roxie/ccd/ccdkey.cpp [hunk 1]
```diff
@@ -1218,8 +1250,11 @@ IDirectReader *InMemoryIndexManager::selectKey(const char *indexSig, ScoredRowFi
     return new InMemoryIndexCursor(this, &thisIndex, baseMap, postFilters, recInfo, translators);
 }
 
-IDirectReader *InMemoryIndexManager::selectKey(ScoredRowFilter &filter, const ITranslatorSet *translators) const
+IDirectReader *InMemoryIndexManager::selectKey(ScoredRowFilter &filter, const ITranslatorSet *translators, IRoxieContextLogger &logctx) const
 {
+    const IKeyTranslator *keyTranslator = translators->queryKeyTranslator(0);;  // any part would do - in-memory requires all actuals to have same layout
```
Comment (ghalliday) [inline roxie/ccd/ccdkey.cpp]: trivial: double ;

File: roxie/ccd/ccdfile.cpp [hunk 2]
```diff
@@ -2127,7 +2134,7 @@ class CResolvedFile : implements IResolvedFileCreator, implements ISDSSubscripti
                         {
                             if (mode == RecordTranslationMode::None)
                                 throw MakeStringException(ROXIE_MISMATCH, "Translatable record layout mismatch detected for file %s, but translation disabled", subname);
-                            keyedTranslator.setown(createKeyTranslator(actual->queryRecordAccessor(true), expected->queryRecordAccessor(true)));
+                            keyedTranslator.setown(createKeyTranslator(actual->queryRecordAccessor(true), expected->queryRecordAccessor(true))); // MORE - should this be outside the if? Should I be checking if it is empty?
```
Comment (ghalliday) [inline roxie/ccd/ccdfile.cpp]: I don't know the answer, I think it is ok, but probably should resolve before this is merged.

**Comment by ghalliday** [comment]
@richardkchapman looks good.  A couple of minor comments.

**Comment by richardkchapman** [comment]
@ghalliday Comments addressed and repushed

File: roxie/ccd/ccdfile.cpp (hunk 2 above)
Comment (richardkchapman) [reply roxie/ccd/ccdfile.cpp]: I THINK inside the IF is ok. Hard to see how the keyed part could need translation if the projected does not. Unless we distinguish expected from projected, and actual happens to match projected but not expected I suppose?

As far as checking it is empty - that will be something that needs doing for index but not for disk cases.

**Comment by richardkchapman** [comment]
Actually not sure I fully addressed one of them - let me check

File: roxie/ccd/ccdactivities.cpp [hunk 3]
```diff
@@ -894,7 +894,7 @@ class CRoxieDiskReadBaseActivity : public CRoxieSlaveActivity, implements IIndex
     {
         unsigned channel = packet->queryHeader().channel;
         unsigned formatCrc = basefactory->getFormatCrc(helper->getFormatCrc());
-        translators.setown(varFileInfo->getTranslators(formatCrc, helper->queryProjectedDiskRecordSize(), helper->queryDiskRecordSize(), basefactory->getEnableFieldTranslation())); // MORE - FormatCRC may be wrong here. Needs to be crc of projected not expected
+        translators.setown(varFileInfo->getTranslators(formatCrc, helper->queryProjectedDiskRecordSize(), helper->queryDiskRecordSize(), basefactory->getEnableFieldTranslation(), true)); // MORE - FormatCRC may be wrong here. Needs to be crc of projected not expected
```
Comment (ghalliday) [inline roxie/ccd/ccdactivities.cpp]: Why is isIndex true?

**Pull request merged**